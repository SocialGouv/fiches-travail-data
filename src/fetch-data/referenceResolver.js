/*
Here we resolve the references : 
Given an article (or a range) and its code (code du travail ou securite sociale), we search for its
actual id in the legi data corpus. 
*/

import { CODE_TRAVAIL, codesFullNames } from "@socialgouv/reference-article";
import find from "unist-util-find";
import visit from "unist-util-visit";

const codes = {};
Object.values(codesFullNames).forEach(({ id }) => {
  const code = require(`@socialgouv/legi-data/data/${id}.json`);
  codes[id] = code;
});

// duplicated in reference Extractor
const rangeMarkers = ["à", "à"];

const CODE_UNKNOWN = { id: "UNDEFINED" };
// shall we use "code du travail" by default ?
const DEFAULT_CODE = CODE_TRAVAIL;

// dumb convert article.data.num as integer for comparison
// each part up to MAX_DEPTH is padded with PAD_LENGTH
const PAD_LENGTH = 5; // left pad numbers to X chars
const MAX_DEPTH = 5; // max number of L432-1-1-1
// padding numbers : 2 -> "0002"
const leftPad = (num) => {
  let padded = "" + num;
  while (padded.length < PAD_LENGTH) {
    padded = "0" + padded;
  }
  return padded;
};
// transform articles into comparable integers
const asInt = (num) => {
  const parts = num
    .replace(/[^\d-]/g, "")
    .split("-")
    .map(leftPad);
  while (parts.length < MAX_DEPTH) {
    parts.push(leftPad(0));
  }
  const int = parseInt(parts.join(""));
  return int;
};

function getLegiDataRange(code, start, end) {
  // check if num is numerically after start. also check LRD prefix
  const isAfterStart = (node) =>
    asInt(node.data.num) >= asInt(start) &&
    node.data.num.charAt(0) === start.charAt(0);

  // check if num is numerically before end. also check LRD prefix
  const isBeforeEnd = (node) =>
    asInt(node.data.num) <= asInt(end) &&
    node.data.num.charAt(0) === end.charAt(0);

  const articles = [];
  visit(code, "article", (node) => {
    if (isAfterStart(node) && isBeforeEnd(node)) {
      articles.push(node);
    }
  });
  return articles;
}

function formatStartEnd(startRaw, endRaw) {
  // we need to identify special case where end ref is relative to start ref (e.g. L. 4733-9 à 11)
  // if there's nothing in common between end and start, we consider being in this special case

  const [startParts, endParts] = [startRaw, endRaw].map((a) =>
    a
      .replace(/\u2011/g, "-")
      .replace()
      .split("-")
      .map((p) => p.trim())
  );

  const letter = startParts[0].slice(0, 1);

  const startNums = Array.from(startParts);
  startNums[0] = startNums[0].replace(/\D/g, "");

  let endNums = Array.from(endParts);
  endNums[0] = endNums[0].replace(/\D/g, "");

  if (endNums.length == 1 && /^\d+$/.test(endParts[0])) {
    const endRange = endNums[0];
    endNums = Array.from(startNums.slice(0, -1));
    endNums.push(endRange);
  }

  return [letter + startNums.join("-"), letter + endNums.join("-")];
}

// in case of a range (like "L. 4733-9 à 4733-11"), we try to identify
// the articles implicitly included within the range
function unravelRange(range) {
  const mark = rangeMarkers.filter((a) => range.text.includes(a))[0];
  const rawParts = range.text.split(mark).map((p) => p.trim());

  const code = range.code ? range.code : DEFAULT_CODE;

  if (rawParts.length == 2 && code != CODE_UNKNOWN) {
    // objective is to identify starting and ending articles (with the legi data correct format)
    // then we can do a legi-data lookup
    const [startRaw, endRaw] = rawParts;
    const [startFMT, endFMT] = formatStartEnd(startRaw, endRaw);

    const unraveled = getLegiDataRange(codes[code.id], startFMT, endFMT).map(
      (a) => {
        const fmt = a.data.num;
        // keep original text for beginning and end
        let text;
        if (startFMT == fmt) {
          text = startRaw;
        } else if (endFMT == fmt) {
          text = endRaw;
        }
        return { ...(text && { text }), code, fmt };
      }
    );

    if (unraveled.length > 0) {
      return unraveled;
    }
  }

  // default in case of error, note that we explicitly set code to unknown
  // in order to identify range errors
  return range.text.split(mark).map((a) => {
    return { code: CODE_UNKNOWN, text: a.trim() };
  });
}

function formatArticle(article) {
  // remove dot and spaces + remove non digit trailing chars + replace unicode dash ‑ to standard -
  return article
    .toUpperCase()
    .replace(".", "")
    .replace(" ", "")
    .replace(/\D*$/, "")
    .replace(/\u2011/g, "-");
}

function resolveReference(ref) {
  let toResolve = [ref];
  if (rangeMarkers.filter((a) => ref.text.includes(a)).length != 0) {
    toResolve = unravelRange(ref);
  }

  return toResolve.map((a) => {
    // use default code if no defined
    const code =
      (a.code == CODE_UNKNOWN) | (a.code == undefined) ? DEFAULT_CODE : a.code;

    if (!a.fmt) a.fmt = formatArticle(a.text);

    if (code && code != CODE_UNKNOWN) {
      const article = find(
        codes[code.id],
        (node) => node.type === "article" && node.data.num === a.fmt
      );
      if (article) {
        a.cid = article.data.cid;
        a.id = article.data.id;
        a.code = code;
      } else {
        // not found in code
        a.code = CODE_UNKNOWN;
      }
    }
    return a;
  });
}

function resolveReferences(refs) {
  const resolvedRefs = refs.map((ref) => resolveReference(ref)).flat();

  const deduplicated = resolvedRefs.reduce((acc, art) => {
    // drop duplicated references
    const existing = acc
      .map((a) => [a.text, a.fmt])
      .flat()
      .filter((v) => v);

    if (!(existing.includes(art.fmt) || existing.includes(art.text))) {
      acc.push(art);
    }
    return acc;
  }, []);

  // group by code
  const grouped = deduplicated.reduce((acc, art) => {
    const { code, ...rawArticle } = art;
    const parsedCode = code ? code : CODE_UNKNOWN;

    if (!Object.keys(acc).includes(parsedCode.id)) {
      acc[parsedCode.id] = { articles: [], name: parsedCode.name };
    }

    acc[parsedCode.id].articles.push(rawArticle);

    return acc;
  }, {});
  return grouped;
}

export { resolveReferences };
