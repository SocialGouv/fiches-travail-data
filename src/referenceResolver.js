/*
Here we resolve the references : 
Given an article and its code (code du travail ou securite sociale), we search for its
actual id in the legi data corpus. 
*/

import find from "unist-util-find";
import { codesFullNames, CODE_TRAVAIL } from "./referenceExtractor";

const codes = {};
Object.values(codesFullNames).forEach(({ id }) => {
  const code = require(`@socialgouv/legi-data/data/${id}.json`);
  codes[id] = code;
});

// duplicated in reference Extractor
const rangeMarkers = ["à", "à"];

function unravelRange(range) {
  // TODO : deal with actual ranges (N to M)
  const mark = rangeMarkers.filter((a) => range.article.includes(a))[0];
  return range.article.split(mark).map((a) => {
    return { article: a.trim(), code: range.code };
  });
}

function formatArticle(article) {
  return article.replace(".", "").replace(" ", "");
}

function resolveReference(ref) {
  let toResolve = [ref];
  if (rangeMarkers.map((a) => ref.article.includes(a))) {
    toResolve = unravelRange(ref);
  }

  return toResolve.map((a) => {
    // by default we try to resolve code du travail
    const codeId = a.code ? a.code.id : CODE_TRAVAIL.id;
    const code = codes[codeId];
    if (code) {
      const formattedArticle = formatArticle(a.article);
      const article = find(
        code,
        (node) => node.type === "article" && node.data.num === formattedArticle
      );
      if (article) {
        a.id = article.data.id;
        a.fmt = formattedArticle;
        // case we guessed code du travail and found article
        if (!a.code) {
          a.code = CODE_TRAVAIL;
        }
      }
    }
    return a;
  });
}

function resolveReferences(refs) {
  return refs.map((ref) => resolveReference(ref)).flat();
}

export { resolveReferences };
