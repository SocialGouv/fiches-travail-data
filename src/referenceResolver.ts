/*
Here we resolve the references :
Given an article (or a range) and its code (code du travail ou securite sociale), we search for its
actual id in the legi data corpus.
*/

import find from "unist-util-find";
import visit from "unist-util-visit";
import { codesFullNames, CODE_TRAVAIL } from "./referenceExtractor";
import { asInt, rangeMarkers } from "./utils";
import type { Reference, ResolvedReferences } from "./types";
// FIXME(rmelisson) : not sure to understand why import Node is causing issue
// eslint-disable-next-line import/no-unresolved
import type { Node } from "unist";

const codes = Object.values(codesFullNames).reduce<{ [id: string]: Node }>(
  (memo, { id }) => {
    const code = require(`@socialgouv/legi-data/data/${id}.json`) as Node;
    return { [id]: code, ...memo };
  },
  {}
);

type CodeArticle = {
  data: {
    num: string;
    id: string;
  };
};

const CODE_UNKNOWN = { id: "UNDEFINED", name: "code undefined" };
// shall we use "code du travail" by default ?
const DEFAULT_CODE = CODE_TRAVAIL;

function getLegiDataRange(
  code: Node,
  start: string,
  end: string
): CodeArticle[] {
  // check if num is numerically after start. also check LRD prefix
  const isAfterStart = (article: CodeArticle): boolean =>
    asInt(article.data.num) >= asInt(start) &&
    article.data.num.charAt(0) === start.charAt(0);

  // check if num is numerically before end. also check LRD prefix
  const isBeforeEnd = (article: CodeArticle): boolean =>
    asInt(article.data.num) <= asInt(end) &&
    article.data.num.charAt(0) === end.charAt(0);

  const articles: CodeArticle[] = [];
  visit(code, "article", (node) => {
    // using untyped tree representation, only option is to cast here
    const article = (node as unknown) as CodeArticle;
    if (isAfterStart(article) && isBeforeEnd(article)) {
      articles.push(article);
    }
  });
  return articles;
}

function formatStartEnd(startRaw: string, endRaw: string): [string, string] {
  // we need to identify special case where end ref is relative to start ref (e.g. L. 4733-9 à 11)
  // if there's nothing in common between end and start, we consider being in this special case

  const [startParts, endParts] = [startRaw, endRaw].map((a) =>
    a
      .replace(/\u2011/g, "-")
      .split("-")
      .map((p) => p.trim())
  );

  const letter = startParts[0].slice(0, 1);

  const startNums = Array.from(startParts);
  startNums[0] = startNums[0].replace(/\D/g, "");

  let endNums = Array.from(endParts);
  endNums[0] = endNums[0].replace(/\D/g, "");

  if (endNums.length == 1 && /^\d+$/.test(endParts[0])) {
    const endRange = endNums[0];
    endNums = Array.from(startNums.slice(0, -1));
    endNums.push(endRange);
  }

  return [letter + startNums.join("-"), letter + endNums.join("-")];
}

type UnravelledReference = {
  text: string;
  fmt: string | undefined;
  code: {
    name: string;
    id: string;
  };
};

// in case of a range (like "L. 4733-9 à 4733-11"), we try to identify
// the articles implicitly included within the range
function unravelRange(range: Reference): UnravelledReference[] {
  const mark = rangeMarkers.filter((a) => range.text.includes(a))[0];
  const rawParts = range.text.split(mark).map((p) => p.trim());

  const code = range.code ? range.code : DEFAULT_CODE;

  if (rawParts.length == 2 && code != CODE_UNKNOWN) {
    // objective is to identify starting and ending articles (with the legi data correct format)
    // then we can do a legi-data lookup
    const [startRaw, endRaw] = rawParts;
    const [startFMT, endFMT] = formatStartEnd(startRaw, endRaw);

    const unraveled = getLegiDataRange(codes[code.id], startFMT, endFMT).map(
      (a) => {
        const fmt = a.data !== undefined ? a.data.num : undefined;
        // keep original text for beginning and end
        let text;
        if (startFMT == fmt) {
          text = startRaw;
        } else if (endFMT == fmt) {
          text = endRaw;
        } else {
          text = range.text + " => " + fmt;
        }
        return { text, fmt, code };
      }
    );

    if (unraveled.length > 0) {
      return unraveled;
    }
  }

  // default in case of error, note that we explicitly set code to unknown
  // in order to identify range errors
  return range.text.split(mark).map((a) => {
    return { text: a.trim(), fmt: undefined, code: CODE_UNKNOWN };
  });
}

function formatArticle(article: string): string {
  // remove dot and spaces + remove non digit trailing chars + replace unicode dash ‑ to standard -
  return article
    .replace(".", "")
    .replace(" ", "")
    .replace(/\D*$/, "")
    .replace(/\u2011/g, "-");
}

function resolveReference(ref: Reference): Reference[] {
  let toResolve: Reference[] = [ref];
  if (rangeMarkers.filter((a) => ref.text.includes(a)).length != 0) {
    toResolve = unravelRange(ref).map((ref) => ref as Reference);
  }

  return toResolve.map((a) => {
    // use default code if no defined
    const code =
      a.code == CODE_UNKNOWN || a.code == undefined ? DEFAULT_CODE : a.code;

    if (!a.fmt) a.fmt = formatArticle(a.text);

    if (code && code != CODE_UNKNOWN) {
      // again we have to cast as nodes are generic
      const article = (find(
        codes[code.id],
        (node: Node) => node.type === "article" && node.data?.num === a.fmt
      ) as unknown) as CodeArticle;
      if (article) {
        a.id = article.data ? article.data.id : "";
        a.code = code;
      } else {
        // not found in code
        a.code = CODE_UNKNOWN;
      }
    }
    return a;
  });
}

function resolveReferences(refs: Reference[]): ResolvedReferences {
  const resolvedRefs = refs.map(resolveReference).flat();

  const deduplicated = resolvedRefs.reduce((acc, art) => {
    // drop duplicated references
    const existing = acc
      .map((a: Reference) => [a.text, a.fmt])
      .flat()
      .filter((v) => v);

    if (!(existing.includes(art.fmt) || existing.includes(art.text))) {
      acc.push(art);
    }
    return acc;
  }, new Array<Reference>());

  // group by code
  const grouped = deduplicated.reduce((acc, art) => {
    const { code, ...rawArticle } = art;
    const parsedCode = code ? code : CODE_UNKNOWN;

    if (!acc.has(parsedCode.id)) {
      acc.set(parsedCode.id, { name: parsedCode.name, articles: [] });
    }
    acc.get(parsedCode.id)?.articles.push(rawArticle);

    return acc;
  }, new Map<string, { name: string; articles: { text: string; fmt: string; id: string }[] }>());
  return Object.fromEntries(grouped);
}

export { resolveReferences };
