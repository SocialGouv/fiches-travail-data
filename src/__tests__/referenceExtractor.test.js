import fs from "fs";
import path from "path";
import { classifyTokens, extractReferences } from "../referenceExtractor";
import { resolveReferences } from "../referenceResolver";

const annotatedTokens = fs
  .readFileSync(path.join(__dirname, "referenceExtractor.test.txt"))
  .toString()
  .split("\n");

const tokens = [];
const labels = [];

annotatedTokens.map((line) => {
  const [t, l] = line.split("\t");
  tokens.push(t);
  labels.push(l);
});

const testCases = [
  {
    input:
      "les modalités fixées par les articles L. 2313-8 et R. 2313-3 à R. 2313-6 du code du travail ainsi que le L. 1251-18",
  },
  {
    input:
      "l’allocation de remplacement pour maternité ou paternité, prévues aux articles L. 613-19 à L.613-19-2 et L. 722-8 à 25 du code de la sécurité sociale, aux articles L. 732-10 à L. 732-12-1 du code rural et à l’article 17 de la loi n° 97-1051 du 18 novembre 1997 d’orientation sur la pêche maritime et les cultures marines",
  },
  {
    input: `Article D212 du code penal et article R413 du code civil`,
  },
  { input: `Article D212` },
  { input: `Article D-212` },
  { input: `Article D.212` },
  { input: `Article D212-3` },
  { input: `Article D-212-4` },
  { input: `Article X*212-4` },
  { input: `Article D. 212-4` },
  { input: `Article D.  212-4` },
  { input: `Article D.212-5` },
  { input: `Article D.212-5-6` },
  { input: `Article D.212-5-6-7` },
  { input: `Article XD212` },
];

test("should extract article tokens in examples", () => {
  expect(
    testCases.map(({ input }) => extractReferences(input))
  ).toMatchSnapshot();
});

test("should success with actual real life set", () => {
  const predictions = classifyTokens(tokens);
  // keep this to easily check prediction errors :
  /*
  const report = tokens.map((t, i) => {
    const label = labels[i];
    const prediction = predictions[i];
    const ok = prediction == label;
    return [
      `${ok ? "\x1b[0m" : "\x1b[31m"}`,
      `${t}\t${label}\t${prediction}\n`
    ];
  });
  console.log(...report.flat());
  */
  expect(predictions).toEqual(labels);
});

it("should find with code for actual real life set", () =>
  expect(extractReferences(tokens.join(" "))).toMatchSnapshot());

it("should resolve example codes", () => {
  const refs = extractReferences(testCases[0].input);
  expect(resolveReferences(refs)).toMatchSnapshot();
});
