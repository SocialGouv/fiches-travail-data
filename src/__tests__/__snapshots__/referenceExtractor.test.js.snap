// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should extract article tokens in examples 1`] = `
Array [
  Array [
    Object {
      "article": "L. 2313-8",
      "code": Object {
        "id": "LEGITEXT000006072050",
        "name": "code du travail",
      },
    },
    Object {
      "article": "R. 2313-3 à R. 2313-6",
      "code": Object {
        "id": "LEGITEXT000006072050",
        "name": "code du travail",
      },
    },
    Object {
      "article": "L. 1251-18",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "L. 613-19 à L.613-19-2",
      "code": Object {
        "id": "LEGITEXT000006073189",
        "name": "code de la sécurité sociale",
      },
    },
    Object {
      "article": "L. 722-8 à 25",
      "code": Object {
        "id": "LEGITEXT000006073189",
        "name": "code de la sécurité sociale",
      },
    },
    Object {
      "article": "L. 732-10 à L. 732-12-1",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D212",
      "code": undefined,
    },
    Object {
      "article": "R413",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D212",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D-212",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D.212",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D212-3",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D-212-4",
      "code": undefined,
    },
  ],
  Array [],
  Array [
    Object {
      "article": "D. 212-4",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D. 212-4",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D.212-5",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D.212-5-6",
      "code": undefined,
    },
  ],
  Array [
    Object {
      "article": "D.212-5-6-7",
      "code": undefined,
    },
  ],
  Array [],
]
`;

exports[`should find with code for actual real life set 1`] = `
Array [
  Object {
    "article": "L. 2313-8",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 2313-3 à R. 2313-6",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 2242-7",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "L. 2242-8",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 2242-2 à R. 2242-11",
    "code": undefined,
  },
  Object {
    "article": "L. 2242-9",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 2242-9 à R. 2242-11",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "L. 135-1",
    "code": undefined,
  },
  Object {
    "article": "D. 1142-2",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "D. 1142-2-1",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "L. 4521-1",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "D. 1142-2",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "D. 1142-2-1",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 1234-9",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "L. 130-1",
    "code": undefined,
  },
  Object {
    "article": "R. 130-1",
    "code": undefined,
  },
  Object {
    "article": "D. 8223-2",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 1263-3",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 1263-4",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 1263-6",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "L. 613-19 à L. 613-19-2",
    "code": undefined,
  },
  Object {
    "article": "L. 722-8 à L. 722-8-3",
    "code": undefined,
  },
  Object {
    "article": "L. 732-10 à L. 732-12-1",
    "code": undefined,
  },
  Object {
    "article": "L. 4412-2",
    "code": undefined,
  },
  Object {
    "article": "R. 4412-97/II",
    "code": undefined,
  },
  Object {
    "article": "L. 2315-24",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "L. 2315-38 à 40",
    "code": undefined,
  },
  Object {
    "article": "L1111-2",
    "code": undefined,
  },
  Object {
    "article": "L. 2314-6",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 2314-6",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "L 2312-36",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  },
  Object {
    "article": "R. 2312-8",
    "code": undefined,
  },
  Object {
    "article": "R. 2312-10 à R. 2312-14",
    "code": undefined,
  },
  Object {
    "article": "R. 2312-9",
    "code": undefined,
  },
  Object {
    "article": "R. 2312-10 à R. 2312-14",
    "code": undefined,
  },
]
`;

exports[`should resolve example codes 1`] = `
Array [
  Object {
    "article": "L. 2313-8",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
    "fmt": "L2313-8",
    "id": "LEGIARTI000036262496",
  },
  Object {
    "article": "R. 2313-3",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
    "fmt": "R2313-3",
    "id": "LEGIARTI000039624940",
  },
  Object {
    "article": "R. 2313-6",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
    "fmt": "R2313-6",
    "id": "LEGIARTI000039624937",
  },
  Object {
    "article": "L. 1251-18",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
    "fmt": "L1251-18",
    "id": "LEGIARTI000006901269",
  },
]
`;
